### PR TITLE
Add fallback content when creating templates according to template hierarchy

### LINKED ISSUE #36648: Block Template creation is using empty content instead of the appropiate fallback
### Description

This was brought to my attention by @carolinan over at https://github.com/WordPress/twentytwentytwo/issues/256

When a theme doesn't provide a template for a specific page (404, Search for example) we will always fallback through the hierarchy and show in the frontend an alternative existing template. In the case of TT2, if the search.html doesn't exist, the frontend will show index.html instead:

Screenshot 2021-11-19 at 13 49 46

This is inconsistent with what the flow of template creation is showing us. When you try to create a new Search template, we see no content at all:

Screenshot 2021-11-19 at 13 49 24

Screenshot 2021-11-19 at 13 49 40

This is inconsistent with what the user sees in the frontend before creating the template and forces them to create something from scratch.

### Step-by-step reproduction instructions

1. Install TT2, check in the frontend what a search results page returns
2. Go to the Site Editor and try to create a Search template
3. The Search template is empty.

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

Close https://github.com/WordPress/gutenberg/issues/36648.

Add fallback content when creating templates according to ["Template Hierarchy"](https://developer.wordpress.org/themes/basics/template-hierarchy/). TBH, I'm not sure if this is the right approach. Feels like it might be better if there's a API to get the fallback content (or just bake it right into the create request?). So that we don't have to maintain a (pseudo-)mapping on front-end, but rather leverage the actual algorithm on back-end to populate the default content.

## How has this been tested?

1. Activate `tt1-blocks` theme
2. Go to Appearance -> Editor
3. Toggle the navigation button and select "Templates"
4. Click "Add New" and select a template to create in the dropdown
5. Notice that the created template has some default content according to template hierarchy

## Types of changes

New feature

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/edit-site/src/components/add-new-template/new-template.js [hunk 1]
```diff
@@ -30,8 +30,40 @@ const DEFAULT_TEMPLATE_SLUGS = [
...
+// @see https://developer.wordpress.org/themes/basics/template-hierarchy/
+const TEMPLATE_HIERARCHY = {
+	'font-page': 'home',
+	home: 'index',
+	'single-post': 'single',
+	single: 'singular',
+	singular: 'index',
+	page: 'singular',
+	archive: 'category',
+	category: 'index',
```
Comment (jameskoster) [inline packages/edit-site/src/components/add-new-template/new-template.js:43]: I think this is slightly incorrect. A new category template should use archive as the fallback. A new archive template should use index as the fallback. 

Edit: We don't currently allow the creation of category templates at all, so perhaps that can be omitted for now?

**Comment by jameskoster** [comment]
I can't comment on whether this is the best approach, but it seems to be working at least :)

File: packages/edit-site/src/components/add-new-template/new-template.js (hunk 1 above)
Comment (MaggieCabrera) [reply packages/edit-site/src/components/add-new-template/new-template.js:43]: and if neither archive or category exist, archive should fall back to index too

**Comment by paaljoachim** [comment]
Downloading the [Gutenberg PR build](https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/). 

In general (testing this PR nor not) what I see is adding any new template the content will look like the index template content. It requires one to delete the content and add the content one needs.
A Search/Archive really should have its own placeholder content. 
Adding specific blocks associated with how a template is to be used. 

Search template would likely add a Heading, a Search block placeholder and a Paragraph block.

Archive template would likely add a Heading, Archive block (I think we have one) and some additional text.

**Comment by noisysocks** [comment]
> TBH, I'm not sure if this is the right approach. Feels like it might be better if there's a API to get the fallback content (or just bake it right into the create request?).

What do you think @ockham?

**Comment by ockham** [comment]
> What do you think @ockham?

:wave: (Sorry, I only noticed the ping now.)

This is an [interesting issue](https://github.com/WordPress/gutenberg/issues/36648) 🤔 My gut reaction is that I'd rather avoid encoding the template hierarchy on the client side. The template hierarchy is a [tricky beast](https://wphierarchy.com/), and I feel it's better to have only one source of truth for it (on the server side, as opposed to duplicating it client-side).

> > TBH, I'm not sure if this is the right approach. Feels like it might be better if there's a API to get the fallback content (or just bake it right into the create request?).

Those sound like better options to me. We already hook into various points to run the template hierarchy resolution algorithm ([e.g.](https://github.com/WordPress/wordpress-develop/pull/1997)), so we might find a point that lends itself to doing this for template creation.

I'll have a look tomorrow unless someone else beats me to it 🙂

**Comment by kevin940726** [comment]
@ockham Awesome! If that's possible then I think we should close this PR and favor that one :)

**Comment by ockham** [comment]
@kevin940726 I looked into this today, but haven't totally gotten to the bottom of it yet.

I was unaware that the "Add New" button on the templates page...

image

...actually creates templates _explicitly_, i.e. by [`POST`ing to the `/wp/v2/templates` endpoint](https://github.com/WordPress/gutenberg/pull/36592/files#diff-ca4c88cde549cde6f09d374e5408049f4e3b53e14978bf6bd01852fd02957382R48-R58) (PR: https://github.com/WordPress/gutenberg/pull/36592).

(FWIW,  this is different from what happens when the user attempts to edit a template that's provided by a theme's block template file -- such as the Archive template in the screenshot above: In this case, the editor loads the template from that file _without_ creating a `wp_template` post. The latter is only created once the user actually modifies and saves the template. This is also the main reason why we're using the "static" `theme//template` IDs -- they allow us to make this process work independently of auto-incremented IDs.)

I'm not totally sure about the explicit creation. There's a point to be made that templates (i.e. `wp_template` posts) should be created the same way -- whether there's a corresponding block template file in the theme or not -- for consistency's sake. This could in particular help with the template hierarchy fallback issue we're trying to solve here.

OTOH, it's fair to acknowledge that the implicit `wp_template` generation is a solution to the problem that it's hard to tell _when_ to create the `wp_template` post. (This is particularly relevant for frontend rendering. A long time ago, we created `wp_template` post auto-drafts whenever a function attempted to read the template -- which might happen upon loading the frontend, potentially long before the user had loaded the template in the site editor. Handling all those auto-drafts proved unwieldy, which is why we came up with the [current solution](https://github.com/WordPress/gutenberg/pull/27910).) However, "Add New" is a very deliberate act of template creation -- so it's perfectly arguable that it calls for explicit template creation.

I'll mull this over a bit more and would like to hear @youknowriad's opinion on this.

I think it's worth to at least experiment with an approach that disables explicit template creation, and makes sure that the [REST API controller's `get_item()`](https://github.com/WordPress/gutenberg/blob/fc8a7891f900aa67a859b28f01aeb4e24fbf9011/lib/compat/wordpress-5.9/class-gutenberg-rest-templates-controller.php#L198-L217) falls back to the next candidate in the template hierarchy, much like we do on the frontend. I'll try this tomorrow.

File: packages/edit-site/src/components/add-new-template/new-template.js [hunk 2]
```diff
@@ -52,13 +86,18 @@ export default function NewTemplate( { postType } ) {
 				slug,
 			} );
 
+			const newTemplateContent =
+				defaultBlockTemplate ??
+				getFallbackTemplateContentFromHierarchy( slug, templates );
```
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-template/new-template.js:91]: It's a bit weird to rely on existing templates to create a fallback template content? Is this how it's supposed to work or should we actually provide a list of default fallbacks?

File: packages/edit-site/src/components/add-new-template/new-template.js [hunk 3]
```diff
@@ -41,6 +73,8 @@ export default function NewTemplate( { postType } ) {
 			defaultTemplateTypes: select(
 				editorStore
 			).__experimentalGetDefaultTemplateTypes(),
+			defaultBlockTemplate: select( editorStore ).getEditorSettings()
```
Comment (youknowriad) [inline packages/edit-site/src/components/add-new-template/new-template.js:76]: This was initially an API suited for "custom templates" and not "hierarchy templates". Is it fine to use it consistently no matter the template? 

Also should we sync the logic here with the fallback done in the template mode? (post editor)

**Review by youknowriad — COMMENTED** [review]
I don't have a strong opinion on whether the fallback should be server side or front-end personally.

File: packages/edit-site/src/components/add-new-template/new-template.js (hunk 2 above)
Comment (jameskoster) [reply packages/edit-site/src/components/add-new-template/new-template.js:91]: The motivation for using existing templates is to ensure that the new template matches the current theme design (IE include the appropriate header / footer template parts and general block structure).

A much simpler approach would be to prescribe a template including the base header / footer template parts, then the user would need to select which one to use whilst creating the template. So a new search template might look like:

Screenshot 2021-12-08 at 10 06 45

The drawback here is that if the theme has a complex layout (maybe the header is actually a sidebar) then this won't match at all.

**Comment by ockham** [comment]
Some early-stage experiment here: https://github.com/WordPress/gutenberg/pull/37258

File: packages/edit-site/src/components/add-new-template/new-template.js (hunk 1 above)
Comment (ockham) [reply packages/edit-site/src/components/add-new-template/new-template.js:43]: BTW I just found a helper that already does something like this: https://github.com/WordPress/gutenberg/blob/2e5af3fc1f2082eabe0f113e728b18d7d3b4703f/packages/edit-site/src/components/navigation-sidebar/navigation-panel/template-hierarchy.js#L21

It's using this `const`: https://github.com/WordPress/gutenberg/blob/556248227df7c6d0fdaab6bdf9f5ff74590fb2b0/packages/edit-site/src/components/navigation-sidebar/navigation-panel/constants.js#L64-L68

We might wanna reuse that (unless we decide to go with https://github.com/WordPress/gutenberg/pull/37258 😊).

**Comment by noisysocks** [comment]
Maybe best to punt this to 5.9.1? I'm not feeling a whole lot of consensus on the approach.

Just FYI, today (Monday) is the last day for this PR to be merged if it is to ship in WP 5.9. I'll begin backporting merged PRs on Tuesday morning Australian Eastern time.

**Comment by jameskoster** [comment]
It would be a real shame if this wasn't included, especially as we had something which did roughly the same job in the past. The original issue really undermines the add-template flows.

**Comment by noisysocks** [comment]
We're out of time unfortunately. I've swapped the labels so that we include this in 5.9.1.

**Comment by kevin940726** [comment]
I think this should be closed, fixing it on frontend side feels clunky and weird. Instead, I think we should focus on the approach in https://github.com/WordPress/gutenberg/pull/37258. @ockham WDYT?

**Comment by ockham** [comment]
> I think this should be closed, fixing it on frontend side feels clunky and weird. Instead, I think we should focus on the approach in #37258. @ockham WDYT?

Hey @kevin940726, thanks for the ping! Well I'm a little biased since #37258 is my PR, so yeah, I would like to see it merged 😊 However, there was some disagreement on the implementation there, and reviews at some point stalled 😅

**Comment by ockham** [comment]
This can probably be closed now that https://github.com/WordPress/gutenberg/pull/42520 has been merged.

**Pull request closed without merging**